sqlmesh/core/macros.py

```diff
@@ -1157,6 +1158,151 @@ def deduplicate(
     return query
 
 
+@macro()
+def date_spine(
+    evaluator: MacroEvaluator,
+    datepart: exp.Expression,
+    start_date: exp.Expression,
+    end_date: exp.Expression,
+) -> t.Union[exp.Alias, exp.Select]:
+    """Returns an aliased QUERY to build a date spine with the given datepart, and range of start_date and end_date. Useful for joining as a date lookup table.
+
+    Args:
+        datepart: The datepart to use for the date spine: day, week, month, year
+        start_date: The start date for the date spine in format YYYY-MM-DD
+        end_date: The end date for the date spine in format YYYY-MM-DD
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlglot.schema import MappingSchema
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "@date_spine('week', '2022-01-20', '2024-12-16')"
+        >>> MacroEvaluator(dialect='duckdb').transform(parse_one(sql)).sql()
+        'SELECT UNNEST(GENERATE_SERIES(CAST(\'2022-01-20\' AS DATE), CAST(\'2024-12-16\' AS DATE), (7 *INTERVAL \'1\' DAY))) AS "date_week"'
+
+    """
+    if datepart.name not in ("day", "week", "month", "year"):
+        raise SQLMeshError(
+            f"Invalid datepart '{datepart.name}'. Expected: 'day', 'week', 'month', or 'year'"
+        )
+
+    try:
+        start_date_obj = datetime.strptime(start_date.name, "%Y-%m-%d").date()
+        end_date_obj = datetime.strptime(end_date.name, "%Y-%m-%d").date()
+    except Exception as e:
+        raise SQLMeshError(
+            f"Invalid date format - start_date and end_date must be in format: YYYY-MM-DD. Error: {e}"
+        )
+
+    if start_date_obj > end_date_obj:
+        raise SQLMeshError(
+            f"Invalid date range - start_date '{start_date.name}' is after end_date '{end_date.name}'."
+        )
+
+    alias_name = f"date_{datepart.name}"
+    start_date_column = exp.cast(start_date, "DATE", dialect=evaluator.dialect)
+    end_date_column = exp.cast(end_date, "DATE", dialect=evaluator.dialect)
+    date_interval = exp.Interval(this=exp.Literal.number(1), unit=datepart.name)
+
+    if evaluator.dialect == "bigquery":
```

Comment:
I don't like this "switching" approach. Why can't we just take a stab at transpiling these array generations?

Reply:
I did, but `generate_date_array` doesn't transpile to any of these other database engines. And you'll notice in bigquery, the alias must happen in the from clause vs. the select clause to work correctly. What do you suggest instead given the above?

Reply:
I'll need to investigate more to answer this, but I'd push for handling this in SQLGlot vs here, because it introduces a pattern that we're generally trying to avoid. I'm pretty confident that we handle many of the nuances already, just need to find the correct AST representation for this concept.

If you have already documented how date spines can be generated in the supported dialects, it would be very helpful if you opened a ticket in SQLGlot. Or we can continue the discussion in Slack and figure out the details -- me and @VaggelisD can help take this to the finish line by improving the transpilation part.

Reply:
Alright we can take this back to slack then, since it sounds like the switching approach is a no go

Reply:
Given https://github.com/tobymao/sqlglot/pull/3899, I think we should be able to do something like this now, more or less:

```python
@macro()
def date_spine(evaluator, start, end, date_part):
    generate_date_array = exp.func(
        "GENERATE_DATE_ARRAY",
        exp.cast(start, "date"),
        exp.cast(end, "date"),
        f"INTERVAL 1 {date_part}"
    )
    exploded = exp.Unnest(expressions=generate_date_array)
    return exp.select(f"date_{date_part}").from_(exploded.as_(f"_dates(date_{date_part}"))
```

Reply:
Note: redshift and mysql use the recursive CTE approach to produce the date spine. We can test / benchmark this and if it turns out too slow there's another idea which is to statically expand the date spine by doing a datediff in python and replacing it with a query that produces static data.